src/main/java/se/liu/ida/hefquin/engine/datastructures/impl/SolutionMappingsHashTable.java

```diff
@@ -1,85 +1,210 @@
+	// TODO: can this be made more efficient?
+	protected final Map<List<Node>, List<SolutionMapping>> map = new HashMap<>();
+	protected final Collection<Var> joinVariables;
+
+	public SolutionMappingsHashTable( final Set<Var> joinVariables ){
+		this.joinVariables = joinVariables;
+	}
+	public SolutionMappingsHashTable( final Var ... vars ) {
+		this.joinVariables = Arrays.asList(vars);
+	}
+
 	@Override
 	public int size() {
-		// TODO Auto-generated method stub
-		return 0;
+		int size = 0;
+		final Iterator<List<SolutionMapping>> li = map.values().iterator();
+		while( li.hasNext() ){
+			size = size + li.next().size();
+		}
+		return size;
 	}
 
 	@Override
 	public boolean isEmpty() {
-		// TODO Auto-generated method stub
-		return false;
+		for ( final List<SolutionMapping> l : map.values() ) {
+			if ( ! l.isEmpty() ) {
+				return false;
+			}
+		}
+		return true;
 	}
 
 	@Override
 	public boolean contains( final Object o ) {
-		// TODO Auto-generated method stub
+		if( !(o instanceof SolutionMapping) ){
+			return false;
+		}
+
+		final SolutionMapping sm = (SolutionMapping) o;
+		for ( final List<SolutionMapping> li : map.values() ) {
+			if ( li.contains(sm) ) {
+				return true;
+			}
+		}
 		return false;
 	}
 
 	@Override
 	public Iterator<SolutionMapping> iterator() {
-		// TODO Auto-generated method stub
-		return null;
+		// The following implementation of this method is inefficient because,
+		// each time it is called, it creates and populates a list of solution
+		// mappings by iterating over the content of this index. However,
+		// for a thread-safe implementation, we may have to live with this.
+		// TODO: is there a more efficient way to implement this method?
+		final List<SolutionMapping> solMap = new ArrayList<>();
+		final Iterator<List<SolutionMapping>> li = map.values().iterator();
+		while(li.hasNext()){
+			solMap.addAll(li.next());
+		}
+		return solMap.iterator();
 	}
 
 	@Override
 	public boolean add( final SolutionMapping e ) {
-		// TODO Auto-generated method stub
-		return false;
+		final List<Node> valKeys = getVarKeys(e);
+
+		List<SolutionMapping> solMapList = map.get(valKeys);
+		if ( solMapList == null) {
+			solMapList = new ArrayList<>();
+			map.put(valKeys, solMapList);
+		}
+
+		solMapList.add(e);
+		return true;
 	}
 
 	@Override
 	public void clear() {
-		// TODO Auto-generated method stub
-		
+		for ( final List<SolutionMapping> l : map.values() ) {
+			l.clear();
+		}
+		map.clear();
 	}
 
 	@Override
 	public Iterable<SolutionMapping> getJoinPartners( final SolutionMapping sm )
 			throws UnsupportedOperationException
 	{
-		// TODO Auto-generated method stub
-		return null;
+		final List<Node> valKeys = getVarKeys(sm);
+		final Iterator<SolutionMapping> matchingSolMaps;
+		if (!valKeys.isEmpty() && valKeys.size() == joinVariables.size()){
+			final List<SolutionMapping> l = map.get(valKeys);
+			matchingSolMaps = (l != null) ? l.iterator() : null;
+		}
+		else if(valKeys.isEmpty()){
```

Comment:
Also, instead of returning an empty valKeys list, getVarKeys should better return null (in cases in which the given solution mapping doesn't have any of the join variables).

Then, after fixing getVarKeys and making it return null in the aforementioned case, you will have to swap the order of the two if-branches here.

Reply:
No, building a hash table without any variable(s) to build it on does not make any sense ;-)

Reply:
I updated the method getVarKeys(). It returns valKeys only when the given solution mapping contains complete join variables. If the solution mapping doesn't contain complete join variables or does not have any of the join variables, this method return null. (b527f7a)